More usable and accessible navigation between View/Edit/Results

jancborchardt opened the pull request:
The icon-only navigation on the top right is not super visible and not understandable, so I worked on some improvements sort-of inspired by our mockups for the new tabs (which we can also eventually use once it’s updated in the component) https://github.com/nextcloud/nextcloud-vue/issues/603

## Before
- Lots of icons
- Some icons show only in some context

https://user-images.githubusercontent.com/925062/195996592-ec09cdde-ca2d-4c1f-95fc-3f1d7be9d979.mp4

## After
- Everything has text
- View/Edit/Results is grouped
- All buttons show always

https://user-images.githubusercontent.com/925062/195996589-2614b829-cd5c-4efd-aedc-ef93ff4d97e7.mp4


## To do
Issues which still need fixing and some of which I need help with @jotoeri @Chartman123 :)
- [x] Sidebar toggle only works in "Edit" mode ("Share" button always switches to "Edit" mode too). This should be possible from everywhere.
- [x] Mobile responsiveness, buttons could be icon-only on mobile
- [x] Test if it works in all views like sharing
- [ ] Loading overlay makes the buttons flash (not essential)

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -26,43 +26,41 @@
+				<template #icon>
+					<IconEye :size="20" />
+				</template>
+				<template>{{ t('forms', 'View') }}</template>
```
```suggestion
				{{ t('forms', 'View') }}
```

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -26,43 +26,41 @@
+				<template #icon>
+					<IconPencil :size="20" />
+				</template>
+				<template>{{ t('forms', 'Edit') }}</template>
```
```suggestion
				{{ t('forms', 'Edit') }}
```

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -26,43 +26,41 @@
+				<template #icon>
+					<IconPoll :size="20" />
+				</template>
+				<template>{{ t('forms', 'Results') }}</template>
```
```suggestion
				{{ t('forms', 'Results') }}
```

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -26,43 +26,41 @@
 			<template #icon>
 				<IconShareVariant :size="20" />
 			</template>
+			<template>{{ t('forms', 'Share form') }}</template>
```
```suggestion
			{{ t('forms', 'Share form') }}
```

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -26,43 +26,41 @@
+			<template>{{ t('forms', 'Share form') }}</template>
 		</NcButton>
-		<NcButton v-if="showSidebarToggle"
+		<NcButton
 			v-tooltip="t('forms', 'Toggle settings')"
```
```suggestion
		<NcButton v-tooltip="t('forms', 'Toggle settings')"
```

Chartman123 reviewed (changes requested):
Some smaller changes to fix the Lint warnings and error. Also I'm not quite sure if we really need the border around the first three buttons

jancborchardt commented:
> Some smaller changes to fix the Lint warnings and error. Also I'm not quite sure if we really need the border around the first three buttons

Thank you for the help, I will look into it! :)
The border is intended to group them, as those 3 belong together, like a "tab bar" of the ones which change the content. Both "Share form" and the sidebar icon are different things, so it’s good to visually separate.

jotoeri commented:
> The border is intended to group them, as those 3 belong together, like a "tab bar" of the ones which change the content. Both "Share form" and the sidebar icon are different things, so it’s good to visually separate.

I agree with that. But i also think we should go with less border radius. Similar to the design you linked in vue. On the results we have it completely without inner radius, but thats indeed probably strange with the inner one of the three.

Chartman123 commented:
> On the results we have it completely without inner radius, but thats indeed probably strange with the inner one of the three.

@jancborchardt I know that this would be some bigger change also to vue component, but would it make sense to switch to the NcCheckboxRadioSwitch think with the alternate button layout and horizontal grouping? At the moment there is no icon slot and I also don't know if that's planned or possible... But would make it a lot more easier for single applications to have such a grouped button element... I already did some experiments with this for the results view switching, but currently it's still very ugly. 😄

jotoeri commented:
@Chartman123 Hmm, be careful not to project anything onto the same component, just because it has a similar base... Components are in fact a reusable version of specialisations of the underlying HTML with specialised behaviour and design.
That does neither forbid us, to use the HTML directly, nor does a single component need to cover the (kind of) generalisation to support all possible use-cases.

Chartman123 commented:
@jotoeri Yes that's true. But if there are any plans at Nextcloud to have these "button groups" standardized for other apps as well, it might be a good thing to place it inside nextcloud-vue and have a re-usable design. I didn't neccessarily mean to "abuse" the NcCheckboxRadioSwitch, but this was my first starting point as the alternate horizontal layout kind of looks like the thing we'd need here.

jancborchardt commented:
> I agree with that. But i also think we should go with less border radius. Similar to the design you linked in vue. On the results we have it completely without inner radius, but thats indeed probably strange with the inner one of the three.

Right, so for now I didn’t go with the design like in Vue because in our case it’s a horizontal button, not a vertical one. I would propose we first go with this design as it’s the least invasive and we use standard buttons there anyway – and then when the tab component is updated, we can use that.

I would also say that NcCheckboxRadioSwitch is not correct semantically here.

What do you think?

jancborchardt commented:
Cool, then I need help with at least that 1 point in the todo :)
> Sidebar toggle only works in "Edit" mode ("Share" button always switches to "Edit" mode too). This should be possible from everywhere.

The sidebar and share button should show in every view as otherwise the tab bar is moving positions.

jotoeri commented:
> > I agree with that. But i also think we should go with less border radius. Similar to the design you linked in vue. On the results we have it completely without inner radius, but thats indeed probably strange with the inner one of the three.
> 
> Right, so for now I didn’t go with the design like in Vue because in our case it’s a horizontal button, not a vertical one. I would propose we first go with this design as it’s the least invasive and we use standard buttons there anyway – and then when the tab component is updated, we can use that.
> 
> I would also say that NcCheckboxRadioSwitch is not correct semantically here.
> 
> What do you think?

Tbh. I'm not fine with that. Why doing half things? It should be just a bit of more CSS (just border-radius?) to get a much better design here.

Chartman123 commented:
@jancborchardt while having a look at the sidebar thing I found a little regression: as you show all the buttons all the time you can also click on a button for the route you're currently at (e.g. click edit button twice). There is an error message then in the browser console...

jancborchardt commented:
> Tbh. I'm not fine with that. Why doing half things? It should be just a bit of more CSS (just border-radius?) to get a much better design here.

@jotoeri I tried it out, and actually the pill radius looks much better and is more fitting to our updated circular design (and the buttons nearby):

## border-radius-pill (like buttons)
https://user-images.githubusercontent.com/925062/195996589-2614b829-cd5c-4efd-aedc-ef93ff4d97e7.mp4

## border-radius-large (like tab bar mockup)
Sorry no video here as my screen recording app seems broken.
![image](https://user-images.githubusercontent.com/925062/197769555-80f51a6e-c745-4482-a4f8-f3cc637e5eb8.png)

@nimishavijay @marcoambrosini what do you think? Ref the tab bar mockups at https://github.com/nextcloud/nextcloud-vue/issues/603

jancborchardt commented:
> @jancborchardt while having a look at the sidebar thing I found a little regression: as you show all the buttons all the time you can also click on a button for the route you're currently at (e.g. click edit button twice). There is an error message then in the browser console...

@Chartman123 Fixed, thanks! :) Still need help with the following if you can:

> Sidebar toggle only works in "Edit" mode ("Share" button always switches to "Edit" mode too). This should be possible from everywhere.

jotoeri commented:
> @jotoeri I tried it out, and actually the pill radius looks much better and is more fitting to our updated circular design (and the buttons nearby). I did commit it so feel free to try it out. :)


Hm, ok. Then i'm with you as a long term design here. 👍

marcoambrosini commented on src/components/TopBar.vue:
```diff
@@ -26,44 +26,44 @@
-				<IconPoll :size="20" />
-			</template>
-		</NcButton>
+		<div class="top-bar-buttons">
```
```suggestion
		<div class="top-bar__buttons">
```

jancborchardt replied:
Fixed :)

marcoambrosini commented on src/components/TopBar.vue:
```diff
@@ -180,6 +180,26 @@ export default {
 	justify-content: flex-end;
 	padding: calc(var(--default-grid-baseline, 4px) * 2);
+
+	&-buttons {
```
```suggestion
	&__buttons {
```

jancborchardt replied:
Fixed :)

marcoambrosini commented:
tbh though I prefer the borderless version, by far

<img width="220" alt="Screenshot 2022-10-26 at 11 14 41" src="https://user-images.githubusercontent.com/26852655/197986608-b6bcd857-b14b-4a6d-bcb0-40f77198791a.png">

jancborchardt commented:
@marcoambrosini borderless just looks not connected at all, it’s not obvious what kind of element it is. What do you think regarding pill style vs rounded borders?

Apart from that, others also do connected, pill-style tab bars (iOS does connected but rounded borders sometimes but I can’t find a proper pattern link):

MS Office on iPad:
![MS office iPad toolbar](https://user-images.githubusercontent.com/925062/198013502-5f80c1d5-3317-434a-a3bc-da29f2a5e9a6.png)

Material design docs website:
![Material tab bar icons](https://user-images.githubusercontent.com/925062/198013512-abcab497-1535-440c-9574-1a21a10f6a28.png)

jancborchardt commented:
Addressed your issues @Chartman123 @marcoambrosini, just as said need help with the fact that the sidebar is only available in "Edit" mode. Possibly could be a follow-up pull request, you tell me. :)

Chartman123 commented:
I'd prefer to have the working sidebar button already in this PR :)

Chartman123 commented:
@jancborchardt I think I fixed the sidebar now. 🙂

jancborchardt commented:
Nice @Chartman123, tested and works! :) Can we then have another review @susnux @hamza221 @jotoeri?

susnux reviewed (commented):
First of all I like this very much! Nice work :+1: 
I tested it and it works for me :smile: 

To me using three different elements in the top bar looks a bit inconsistent or messy (meaning it does not look clean like the other UI).
So there is the 3-button array, then there is the "share form" button and next to it is an other icon-only sidebar-toggle.
Maybe one could integrate that share button with the 3-button array or maybe even better with the "form settings"-sidebar-toggle.

But that is only my personal preference, you might ignore that :wink: 


A more important thing is the second ToDo, on mobile it is unusable, this is how it looks for me:
![Menubar overlows on the left side, it is unusable](https://user-images.githubusercontent.com/1855448/202278140-4d570368-8704-4670-92f0-5f7ad32430ff.png)

Chartman123 commented:
> on mobile it is unusable,

Should be fixed now, I added the isMobile mixin 🙂

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -25,42 +25,50 @@
+				<template #icon>
+					<IconEye :size="20" />
+				</template>
+				<template v-if="!isMobile">
+					{{ t('forms', 'View') }}
+				</template>
```
@jancborchardt do we need the aria-labels again now that the text is shown only if !isMobile?

jancborchardt replied:
Ah yes you are absolutely right! On isMobile we should use aria-label then.

Chartman123 replied:
@jancborchardt Do you want to add this to the PR? I'm currently quite busy...

jancborchardt replied:
Yep, I will try with these changes and also the ones from @jotoeri that I can manage – I might have to ask you (or @hamza221 ;) how to do the stuff that go beyond my frontend skills. :D

jancborchardt replied:
So I would use `:aria-label="isMobile ? t('forms', 'View form') : ''"` but then we end up with an empty `aria-label=""` on desktop, which is dirty. Is there a way to get around this?
(Tbh instead of the isMobile mixin I would normally do this via CSS media queries?)

Chartman123 replied:
@jancborchardt I'm also not a vuejs pro 😆 I just tried the following and it seems to do the job: `:aria-label="isMobile ? t('forms', 'View form'):null"`

jancborchardt replied:
Niiice, thanks a lot @Chartman123 that totally worked indeed. :) Fixed!

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -25,42 +25,50 @@
+				<template #icon>
+					<IconPencil :size="20" />
+				</template>
+				<template v-if="!isMobile">
+					{{ t('forms', 'Edit') }}
+				</template>
```
Same here

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -25,42 +25,50 @@
+				<template #icon>
+					<IconPoll :size="20" />
+				</template>
+				<template v-if="!isMobile">
+					{{ t('forms', 'Results') }}
+				</template>
```
Same here

Chartman123 commented on src/components/TopBar.vue:
```diff
@@ -25,42 +25,50 @@
 			<template #icon>
 				<IconShareVariant :size="20" />
 			</template>
+			<template v-if="!isMobile">
+				{{ t('forms', 'Share form') }}
+			</template>
```
Same here

jancborchardt replied:
Yup, agree. Also for simplicity, the visual text here can be called just "Share".

For aria-label, "Share form" is better.

Chartman123 replied:
"Share" would be fine for me

Chartman123 reviewed (commented):
Question regarding aria-labels

jotoeri commented on src/Forms.vue:
```diff
@@ -49,6 +49,7 @@
 					:key="form.id"
 					:form="form"
 					:read-only="true"
+					@open-sharing="openSharing"
```
Seems to be a rest of some testing. That functionality will not be available currently if i see that right. Not best practice for a PR, but i'm ok with leaving it here, since we will most probably use it somewhen.

jotoeri commented on src/Forms.vue:
```diff
@@ -92,7 +93,7 @@
 				:sidebar-opened.sync="sidebarOpened"
 				@open-sharing="openSharing" />
-			<router-view v-if="!selectedForm.partial"
+			<router-view v-if="!selectedForm.partial && selectedForm.permissions.includes(PERMISSION_TYPES.PERMISSION_EDIT)"
```
computed `canEdit` for nice code?

jotoeri commented on src/Forms.vue:
```diff
@@ -236,7 +237,8 @@ export default {
 		 */
 		openSharing(hash) {
-			if (hash !== this.routeHash || this.$route.name !== 'edit') {
+			// if (hash !== this.routeHash || this.$route.name !== 'edit') {
```
To remove

Chartman123 replied:
Ah yes, I had that on my list, but forgot to remove it yesterday when I squashed the fixup commits...

jancborchardt replied:
Removed :)

jotoeri commented on src/components/TopBar.vue:
```diff
@@ -25,42 +25,50 @@
-				<IconPoll :size="20" />
-			</template>
-		</NcButton>
+		<div class="top-bar__buttons">
```
Just naming wise - There are more buttons in the toolbar. Probably more a `top-bar__view-select` or sth like that?

jancborchardt replied:
You’re right, fixed!

jotoeri commented on src/components/TopBar.vue:
```diff
@@ -180,6 +189,24 @@ export default {
+		border: 2px solid var(--color-border);
+		border-radius: var(--border-radius-pill);
+
+		.button-vue {
+
+			&:disabled {
+				opacity: initial;
+				filter: initial;
+			}
+		}
```
Reasonably-wise it is strange to disable the buttons but overriding the look to not-disabled.
Maybe it would be more reasonable to conditionally stop the problematic router-push (or whatever it is) instead of disabling the buttons?

jancborchardt replied:
You’re right. I removed disabling the controls and the styling as this is also not done in other solutions.

There is a Vue router issue (`NavigationDuplicated Navigating to current location`) which is out of my expertise though. I found this for further reference, maybe @Chartman123 @jotoeri you can have a look?
- https://github.com/vuejs/vue-router/issues/2881#issuecomment-520554378
- https://stackoverflow.com/questions/57837758/navigationduplicated-navigating-to-current-location-search-is-not-allowed

jotoeri replied:
I'd just surround the router push in `showEdit`... with the if-statement, that you basically had before to disable the button. That's also what the second part of the accepted answer on your linked stackoverflow thread wants to do. That cathes the error before it gets produced. The other mentioned solution would produce the error and catch it afterwards. 🤷‍♂️

Chartman123 replied:
Should be fixed with my latest commit 👍🏻

jotoeri replied:
Uhhhm, i'm sorry, but i don't think this is a good way:
- Normally one would try more to get from sync calls to async calls, since the latter does not block the whole process. This is also most probably, why the Router process changed from sync to async as mentioned in the issue. The now implemented way does just the oposite. 🙈 
- Quite complex code, that one will wonder in 2 years, why this is done
- Calls a method, of which we exactly know, that it will throw an error, just to catch the error then.

I see, that it was for sure nice to implement that and see, that it works, but i'd rather go with the simple if-version i mentioned above. 😉 Sorry 🤷‍♂️

Chartman123 replied:
I didn't read the second link as I tried to remove the error after reading the first link ;) But yes, makes more sense. So I'll revert my commit then 👍🏻

jotoeri commented on src/views/Results.vue:
```diff
@@ -183,6 +186,13 @@ export default {
 
 	mixins: [ViewsMixin],
 
+	props: {
+		sidebarOpened: {
+			type: Boolean,
+			required: true,
+		},
+	},
+
```
This is now used on all three views. So it should better go to the ViewsMixin?

jancborchardt replied:
Fixed - I needed to remove the `required` attribute as otherwise Sidebar.vue had an error of a missing required prop – let me know if it should be done otherwise. (@hamza221 said we could also pass `:sidebarOpened` to the view in Forms.vue, but with what value then – it’s the sidebar itself. ;)

jotoeri replied:
Uhm, the sidebar anyways takes a required value `opened`. If you replace that one and its usage with `sidebarOpened` it should be good. :)

jotoeri replied:
Unfortunately i couldn't test it now, but i think the added commit should be it already.

jotoeri commented on src/views/Results.vue:
```diff
@@ -324,6 +334,10 @@ export default {
 
 			return submissions
 		},
+
+		onSidebarChange(newState) {
+			this.$emit('update:sidebarOpened', newState)
+		},
```
Same here - ViewsMixin?

jotoeri reviewed (commented):
Didn't test, just had a short look into the code. :relaxed:

jancborchardt commented:
@chartman123 @jotoeri fixed all the issues except the `Duplicated Navigation` thing at https://github.com/nextcloud/forms/pull/1381#discussion_r1037452929 – I either need some pointers, or you are welcome to fix it, or we accept the error (not sure if that should be done). Thanks! :)

Chartman123 commented:
@jancborchardt Nice 🎉  I think that we should clean up the commits a little bit. In my opinion we can squash all commits except the last one into a big one. What do you think?

jotoeri commented:
Could just test it, the Sidebar thing works. Just two more things:
- We shouldn't show the View Switch, if only 'view' is available. Having a shared form with no permissions than submitting, it currently shows the View-Topbar Part.
- @jancborchardt Can we get the border somewhat inside of the view-select? Currently we have a height of 44px+2*2px border.  Can we just fix the height of the view-select to 44px?

Chartman123 commented:
> * We shouldn't show the View Switch, if only 'view' is available. Having a shared form with no permissions than submitting, it currently shows the View-Topbar Part.

@jotoeri fixed it with my last commit

jotoeri commented:
> @jotoeri fixed it with my last commit

I shifted it once more, since the viewselect border was still visible.
Also added the fixed height now for the view-select.

https://github.com/nextcloud/forms/pull/1381/commits/71a6837d55676269e739c2a89d8e3b628432cdf5

|before|after|
|---|---|
|![grafik](https://user-images.githubusercontent.com/47433654/205381892-7e697263-07dd-44eb-8c19-6784ba94dca7.png)|![grafik](https://user-images.githubusercontent.com/47433654/205381945-34cb691e-d1d6-4488-a9c0-b7238c8158c3.png)|

jotoeri reviewed (approved):
Fine by me now
